File: ObjcExample/ObjcExample/Demo/GKPageSmoothView/Custom/GKSmoothCustomViewController.m
//
//  GKSmoothCustomViewController.m
//  ObjcExample
//
//  Created by QuintGao on 2025/3/18.
//

#import "GKSmoothCustomViewController.h"
#import <GKPageSmoothView/GKPageSmoothView.h>
#import "GKCustomSegmentedView.h"
#import "GKBaseListViewController.h"

@interface GKSmoothCustomViewController ()<GKPageSmoothViewDataSource, GKPageSmoothViewDelegate, GKCustomSegmentedViewDelegate>

@property (nonatomic, strong) GKPageSmoothView *smoothView;

@property (nonatomic, strong) UIImageView *headerView;

@property (nonatomic, strong) GKCustomSegmentedView *segmentedView;

@end

@implementation GKSmoothCustomViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_navTitle = @"自定义Segmented";
    self.gk_navLineHidden = YES;
    self.gk_navTitleColor = [UIColor whiteColor];
    self.gk_navTitleFont = [UIFont boldSystemFontOfSize:18.0f];
    self.gk_navBackgroundColor = [UIColor clearColor];
    
    [self.view addSubview:self.smoothView];
    [self.smoothView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    [self.smoothView reloadData];
}

#pragma mark - GKPageSmoothViewDataSource
- (UIView *)headerViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.headerView;
}

- (UIView *)segmentedViewInSmoothView:(GKPageSmoothView *)smoothView {
    return self.segmentedView;
}

- (NSInteger)numberOfListsInSmoothView:(GKPageSmoothView *)smoothView {
    return 3;
}

- (id<GKPageSmoothListViewDelegate>)smoothView:(GKPageSmoothView *)smoothView initListAtIndex:(NSInteger)index {
    GKBaseListViewController *listVC = [[GKBaseListViewController alloc] initWithListType:index];
    return listVC;
}

#pragma mark - GKPageSmoothViewDelegate
- (void)smoothView:(GKPageSmoothView *)smoothView listDidAppearAtIndex:(NSInteger)index {
    [self.segmentedView scrollSelectWithIndex:index];
}

#pragma mark - GKCustomSegmentedViewDelegate
- (void)didClickSelectAtIndex:(NSInteger)index {
    [self.smoothView selectListWithIndex:index animated:YES];
}

#pragma mark - lazy
- (GKPageSmoothView *)smoothView {
    if (!_smoothView) {
        _smoothView = [[GKPageSmoothView alloc] initWithDataSource:self];
        _smoothView.delegate = self;
        _smoothView.listCollectionView.gk_openGestureHandle = YES;
    }
    return _smoothView;
}

- (UIImageView *)headerView {
    if (!_headerView) {
        _headerView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, kBaseHeaderHeight)];
        _headerView.image = [UIImage imageNamed:@"test"];
    }
    return _headerView;
}

- (GKCustomSegmentedView *)segmentedView {
    if (!_segmentedView) {
        _segmentedView = [[GKCustomSegmentedView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, kBaseSegmentHeight)];
        _segmentedView.delegate = self;
    }
    return _segmentedView;
}

@end

File: SwiftExample/SwiftExample/Demo/GKPageSmoothView/Custom/GKSmoothCustomViewController.swift
//
//  GKSmoothCustomViewController.swift
//  SwiftExample
//
//  Created by QuintGao on 2025/3/18.
//

import UIKit
import GKPageSmoothView

class GKSmoothCustomViewController: GKDemoBaseViewController {

    lazy var smoothView: GKPageSmoothView = {
        let smoothView = GKPageSmoothView(dataSource: self)
        smoothView.delegate = self;
        smoothView.ceilPointHeight = kNavBar_Height
        smoothView.listCollectionView.gk_openGestureHandle = true
        return smoothView
    }()
    
    lazy var headerView: UIImageView = {
        let imageView = UIImageView(frame: CGRectMake(0, 0, view.frame.width, kBaseHeaderHeight))
        imageView.image = UIImage(named: "test")
        return imageView
    }()
    
    lazy var segmentedView: GKCustomSegmentedView = {
        let view = GKCustomSegmentedView(frame: CGRectMake(0, 0, view.frame.width, kBaseSegmentHeight))
        view.delegate = self;
        return view
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        gk_navTitleColor = .white
        gk_navTitleFont = .boldSystemFont(ofSize: 18)
        gk_navBackgroundColor = .clear
        gk_navLineHidden = true
        gk_statusBarStyle = .lightContent
        gk_navTitle = "自定义Segmented"
        
        view.addSubview(smoothView)
        smoothView.snp.makeConstraints {
            $0.edges.equalTo(self.view)
        }
        DispatchQueue.main.asyncAfter(wallDeadline: .now() + 0.1) {
            self.smoothView.reloadData()
        }
    }
}

extension GKSmoothCustomViewController: GKPageSmoothViewDataSource, GKPageSmoothViewDelegate {
    func headerView(in smoothView: GKPageSmoothView) -> UIView {
        headerView
    }
    
    func segmentedView(in smoothView: GKPageSmoothView) -> UIView {
        segmentedView
    }
    
    func numberOfLists(in smoothView: GKPageSmoothView) -> Int {
        3
    }
    
    func smoothView(_ smoothView: GKPageSmoothView, initListAtIndex index: Int) -> any GKPageSmoothListViewDelegate {
        GKBaseListViewController(listType: GKBaseListType(rawValue: index) ?? .UITableView)
    }
    
    func smoothViewListDidAppear(_ smoothView: GKPageSmoothView, at index: Int) {
        segmentedView.scrollSelect(with: index)
    }
}

extension GKSmoothCustomViewController: GKCustomSegmentedViewDelegate {
    func didClickSelect(at index: Int) {
        smoothView.selectList(with: index, animated: true)
    }
}
